File: ExpandableAllocator/Allocator.fs
namespace ExpandableAllocator

#nowarn "25" "51"

open Foreign
open System
open System.IO

/// Represents the authorization of an allocated memory block.
[<Flags>]
type Protection = Read    = 0x1
                | Write   = 0x2
                | Execute = 0x4

                | ReadWrite        = 0x3
                | ReadExecute      = 0x5
                | ReadWriteExecute = 0x7

[<AutoOpen>]
module internal ProtectionUtils =
    type Protection with
        member prot.WindowsValue =
            match prot with
            | rwx when rwx = (Protection.Read ||| Protection.Write ||| Protection.Execute) -> 0x40
            | rx  when rx  = (Protection.Read ||| Protection.Execute) -> 0x20
            | rw  when rw  = (Protection.Read ||| Protection.Write) -> 0x04

            | Protection.Read    -> 0x02
            | Protection.Execute -> 0x10

            | _ -> raise(InvalidOperationException("Cannot protect using the provided flags."))

        member prot.UnixValue = int prot


/// Defines an allocator whose memory never moves and may grow until a given amount.
type Allocator private(start: nativeint, maxSize: nativeint, protection: Protection) =
    let mutable actualSize = nativeint 0
    let mutable prot = protection

    /// Gets a pointer to the start of the allocated memory region.
    member __.Start = start

    /// Gets or sets the protection of the reserved region.
    member __.Protection
        with get() = prot
        and  set(value) = if value <> prot then
                              let os = OS.Current

                              let ok = match os with
                                       | OS.Windows ->
                                           let mutable old = 0

                                           Windows.VirtualProtect(start, actualSize, value.WindowsValue, &&old)
                                          
                                       | _ ->
                                           let mprotect = match os with
                                                          | OS.Unix -> Unix.mprotect
                                                          | OS.OSX -> OSX.mprotect

                                           mprotect(start, actualSize, value.UnixValue) = 0
                              
                              if ok then
                                  prot <- value
                              else
                                  raise(IOException("Could not change memory protection."))


    /// Gets the maximum size of memory that can be allocated.
    member __.MaximumSize = maxSize

    /// Gets or sets the actual size of the allocated memory.
    member this.ActualSize
        with get() = actualSize
        and  set(value) = if this.TryReserve(value) |> not then
                              raise(IOException("Could not change actual size of allocated memory."))

    /// Attempts to reserve the given amount of memory.
    member __.TryReserve(size: nativeint) =
        if size <= actualSize then
            true
        elif size > maxSize then
            false
        else
            let os = OS.Current

            match os with
            | OS.Windows ->
                match Windows.VirtualAlloc(start, size, 0x00001000 (* MEM_COMMIT *), prot.WindowsValue) with
                | zero when zero = IntPtr.Zero -> false
                | _ -> actualSize <- size; true

            | _ ->
                let mprotect = match os with
                               | OS.Unix -> Unix.mprotect
                               | OS.OSX -> OSX.mprotect

                match mprotect(start, size, int prot) with
                | 0 -> actualSize <- size; true
                | _ -> false
            

    interface IDisposable with
        /// Disposes the allocator, freeing all its reserved and allocated memory.
        member __.Dispose() =
            match OS.Current with
            | OS.Unix    -> Unix.munmap(start, actualSize) |> ignore
            | OS.OSX     -> OSX.munmap(start, actualSize)  |> ignore
            | OS.Windows -> Windows.VirtualFree(start, actualSize, 0x8000 (* MEM_RELEASE *)) |> ignore


    /// <summary>
    ///   Attempts to create a new allocator, given the protection of the memory
    ///   region and its maximum size.
    /// </summary>
    /// <param name="protection">
    ///   The protection to apply to the allocated region.
    /// </param>
    /// <param name="maxSize">
    ///   The maximum size that can be allocated through the allocator.
    ///   This size may be greater than the available RAM.
    /// </param>
    /// <returns>
    ///   On success, an allocator that can use the reserved memory.
    /// </returns>
    static member TryCreate(protection: Protection, maxSize: nativeint) =
        if maxSize = IntPtr.Zero then
            None
        else
            let os = Foreign.OS.Current

            match os with
            | OS.Windows ->
                match Windows.VirtualAlloc(IntPtr.Zero, maxSize, 0x00002000 (* MEM_RESERVE *), protection.WindowsValue) with
                | zero when zero = IntPtr.Zero -> None
                | addr -> Some(new Allocator(addr, maxSize, protection))

            | _ ->
                let mmap = match os with
                           | OS.Unix -> Unix.mmap
                           | OS.OSX -> OSX.mmap

                match mmap(IntPtr.Zero, maxSize, 0x0 (* PROT_NONE *), 0, 0, IntPtr.Zero) with
                | zero when zero = IntPtr.Zero -> None
                | addr -> Some(new Allocator(addr, maxSize, protection))

    /// <summary>
    ///   Creates a new allocator, given the protection of the memory
    ///   region and its maximum size.
    /// </summary>
    /// <param name="protection">
    ///   The protection to apply to the allocated region.
    /// </param>
    /// <param name="maxSize">
    ///   The maximum size that can be allocated through the allocator.
    ///   This size may be greater than the available RAM.
    /// </param>
    /// <returns>
    ///   An allocator that can use the reserved memory.
    /// </returns>
    /// <exception cref="NullReferenceException" />
    static member Create(protection: Protection, maxSize: nativeint) =
        Allocator.TryCreate(protection, maxSize).Value

File: ExpandableAllocator/Allocator.cs
using System;
using System.IO;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

namespace ExpandableAllocator
{
    /// <summary>
    ///   Represents the protection of a reserved or commited memory block.
    /// </summary>
    public enum Protection
    {
        /// <summary>Permission to read.</summary>
        Read = 0x1,
        /// <summary>Permission to write.</summary>
        Write = 0x2,
        /// <summary>Permission to execute.</summary>
        Execute = 0x4,

        /// <summary>Permission to read and write.</summary>
        ReadWrite = Read | Write,
        /// <summary>Permission to read and execute.</summary>
        ReadExecute = Read | Execute,
        /// <summary>Permission to read, write and execute.</summary>
        ReadWriteExecute = Read | Write | Execute
    }

    internal static class ProtectionHelpers
    {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static int GetWindowsValue(this Protection prot)
        {
            switch (prot)
            {
                case Protection.Read: return 0x02;
                case Protection.Execute: return 0x10;
                case Protection.ReadWrite: return 0x04;
                case Protection.ReadExecute: return 0x20;
                case Protection.ReadWriteExecute: return 0x40;

                default:
                    throw new ArgumentOutOfRangeException(nameof(prot), "Invalid protection.");
            }
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static int GetUnixValue(this Protection prot) => (int) prot;
    }

    /// <summary>
    ///   Defines an allocator whose memory never moves and may grow until a given amount.
    /// </summary>
    public sealed class Allocator : IDisposable
    {
        private IntPtr size;
        private Protection prot;

        private Allocator(IntPtr addr, IntPtr maxSize, Protection protection)
        {
            Address = addr;
            MaximumSize = maxSize;

            prot = protection;
        }

        /// <summary>
        ///   Gets a pointer to the start of the allocated memory region.
        /// </summary>
        public IntPtr Address { get; }

        /// <summary>
        ///   Gets the maximum size of memory that can be allocated.
        /// </summary>
        public IntPtr MaximumSize { get; }

        /// <summary>
        ///   Gets or sets the protection of the reserved region.
        /// </summary>
        public Protection Protection
        {
            get => prot;
            
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            set
            {
                bool ok;
                OS os = Foreign.CurrentOS;

                switch (os)
                {
                    case OS.Windows:
                        ok = Foreign.Windows.VirtualProtect(Address, size, value.GetWindowsValue(), out int _);
                        break;

                    case OS.OSX:
                        ok = Foreign.OSX.mprotect(Address, size, value.GetUnixValue()) == 0;
                        break;

                    case OS.Unix:
                        ok = Foreign.Unix.mprotect(Address, size, value.GetUnixValue()) == 0;
                        break;

                    default:
                        // This cannot happen.
                        ok = false;
                        break;
                }

                if (ok)
                    prot = value;
                else
                    throw new IOException($"Unable to change memory protection: {Marshal.GetLastWin32Error()}.");
            }
        }

        /// <summary>
        ///   Gets or sets the actual size of the allocated memory.
        /// </summary>
        public IntPtr ActualSize
        {
            get => size;
            
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            set
            {
                if (!TryReserve(value))
                    throw new IOException($"Unable to change actual size of allocated memory: {Marshal.GetLastWin32Error()}.");
            }
        }

        /// <summary>
        ///   Attempts to reserve the given amount of memory.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public bool TryReserve(long size) => TryReserve(new IntPtr(size));

        /// <summary>
        ///   Attempts to reserve the given amount of memory.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public bool TryReserve(IntPtr size)
        {
            IntPtr actualSize = this.size;

            long sizeL = size.ToInt64();

            if (sizeL <= actualSize.ToInt64())
                return true;
            if (sizeL > MaximumSize.ToInt64())
                return false;

            OS os = Foreign.CurrentOS;

            switch (os)
            {
                case OS.Windows:
                    IntPtr addr = Foreign.Windows.VirtualAlloc(Address, size, 0x00001000 /* MEM_RESERVE */, prot.GetWindowsValue());
                    
                    if (addr == IntPtr.Zero)
                        return false;

                    break;

                default:
                    int status = os == OS.OSX
                               ? Foreign.OSX.mprotect(Address, size, prot.GetUnixValue())
                               : Foreign.Unix.mprotect(Address, size, prot.GetUnixValue());

                    if (status != 0)
                        return false;

                    break;
            }

            this.size = size;
            return true;
        }

        /// <summary>
        ///   Disposes of the allocator, freeing all its reserved and allocated memory.
        /// </summary>
        public void Dispose()
        {
            switch (Foreign.CurrentOS)
            {
                case OS.Windows:
                    Foreign.Windows.VirtualFree(Address, size, 0x8000 /* MEM_RELEASE */);
                    break;
                case OS.Unix:
                    Foreign.Unix.munmap(Address, size);
                    break;
                case OS.OSX:
                    Foreign.OSX.munmap(Address, size);
                    break;
            }
        }


        /// <summary>
        ///   Attempts to create a new allocator, given the protection of the memory
        ///   region and its maximum size.
        /// </summary>
        /// <param name="protection">
        ///   The protection to apply to the allocated region.
        /// </param>
        /// <param name="maxSize">
        ///   The maximum size that can be allocated through the allocator.
        ///   This size may be greater than the available RAM.
        /// </param>
        /// <param name="allocator">
        ///   On success, an allocator that can be used to reserve memory.
        /// </param>
        /// <returns>
        ///   Whether an allocator was successfully created.
        /// </returns>
        public static bool TryCreate(Protection protection, IntPtr maxSize, out Allocator allocator)
        {
            allocator = null;

            if (maxSize == IntPtr.Zero)
                return false;

            OS os = Foreign.CurrentOS;
            IntPtr addr;

            switch (os)
            {
                case OS.Windows:
                    addr = Foreign.Windows.VirtualAlloc(IntPtr.Zero, maxSize,
                                                        0x00002000 /* MEM_RESERVE */,
                                                        protection.GetWindowsValue());
                    break;

                default:
                    addr = os == OS.OSX
                         ? Foreign.OSX.mmap(IntPtr.Zero, maxSize, 0x0 /* PROT_NONE */, 0x22 /* MAP_ANONYMOUS | MAP_PRIVATE */, -1, IntPtr.Zero)
                         : Foreign.Unix.mmap(IntPtr.Zero, maxSize, 0x0, 0x22, -1, IntPtr.Zero);
                    break;
            }

            if (addr == IntPtr.Zero)
                return false;

            allocator = new Allocator(addr, maxSize, protection);
            return true;
        }

        /// <summary>
        ///   Attempts to create a new allocator, given the protection of the memory
        ///   region and its maximum size.
        /// </summary>
        /// <param name="protection">
        ///   The protection to apply to the allocated region.
        /// </param>
        /// <param name="maxSize">
        ///   The maximum size that can be allocated through the allocator.
        ///   This size may be greater than the available RAM.
        /// </param>
        /// <param name="allocator">
        ///   On success, an allocator that can be used to reserve memory.
        /// </param>
        /// <returns>
        ///   Whether an allocator was successfully created.
        /// </returns>
        public static bool TryCreate(Protection protection, long maxSize, out Allocator allocator)
            => TryCreate(protection, new IntPtr(maxSize), out allocator);

        /// <summary>
        ///   Creates a new allocator, given the protection of the memory
        ///   region and its maximum size.
        /// </summary>
        /// <param name="protection">
        ///   The protection to apply to the allocated region.
        /// </param>
        /// <param name="maxSize">
        ///   The maximum size that can be allocated through the allocator.
        ///   This size may be greater than the available RAM.
        /// </param>
        /// <returns>
        ///   An allocator that can be used the reserved memory.
        /// </returns>
        /// <exception cref="IOException"/>
        public static Allocator Create(Protection protection, IntPtr maxSize)
        {
            if (TryCreate(protection, maxSize, out Allocator allocator))
                return allocator;
            
            throw new IOException($"Could not create an allocator: {Marshal.GetLastWin32Error()}.");
        }

        /// <summary>
        ///   Creates a new allocator, given the protection of the memory
        ///   region and its maximum size.
        /// </summary>
        /// <param name="protection">
        ///   The protection to apply to the allocated region.
        /// </param>
        /// <param name="maxSize">
        ///   The maximum size that can be allocated through the allocator.
        ///   This size may be greater than the available RAM.
        /// </param>
        /// <returns>
        ///   An allocator that can be used the reserved memory.
        /// </returns>
        /// <exception cref="IOException"/>
        public static Allocator Create(Protection protection, long maxSize)
            => Create(protection, new IntPtr(maxSize));
    }
}